server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -944,6 +945,43 @@ static void validateDataStreamOptions(
         }
     }
 
+    /**
+     * Maybe runs {@link #validateDataStreamsStillReferenced} if it looks like the new composite template could change data stream coverage.
+     */
+    private static void maybeValidateDataStreamsStillReferenced(
+        ProjectMetadata project,
+        String templateName,
+        ComposableIndexTemplate newTemplate
+    ) {
+        final ComposableIndexTemplate existing = project.templatesV2().get(templateName);
+        final Settings existingSettings = Optional.ofNullable(existing)
+            .map(ComposableIndexTemplate::template)
+            .map(Template::settings)
+            .orElse(Settings.EMPTY);
+        final Settings newSettings = Optional.ofNullable(newTemplate)
+            .map(ComposableIndexTemplate::template)
+            .map(Template::settings)
+            .orElse(Settings.EMPTY);
+        // We check whether anything relevant has changed that could affect data stream coverage and return early if not.
+        // These checks are based on the implementation of findV2Template and the data stream template check in this method.
+        // If we're adding a new template, we do the full check in case this template's priority changes coverage.
+        if (existing != null
```

Comment:
This a nit, but it's throwing me off: the use of `Objects.equals` here is inconsistent. I'd rather it's either used exactly and only where it is 100% necessary for the code to be correct, or that it's just used for all the stanzas without thinking. As it is right now, the `Objects.equals` around the two `!= null` checks seems unnecessary.

If you're truly neutral on the `only-where-necessary` versus `always` front, I *slightly* prefer `always`, because then instead of a mix of `a.equals(b)` and `a == b` and `Object.equals(a, b)`, we'd just have only `Object.equals(a,  b)` four times.

What do you think?

Reply:
This crossed my mind too! I preferred `Object.equals` for the `!= null` checks because I felt like `(existingTemplate.getDataStreamTemplate() != null) == (newTemplate.getDataStreamTemplate() != null)` would be annoying to read and I didn't know how our spotless configuration would format it. Although the boolean boxing is unfortunate, I figured it wasn't that problematic. I agree on the consistency, though. I'll push a commit to sync them.